### PR TITLE
feat(S2S): implement S2S authentication as Express middleware

### PR DESCRIPTION
**Pull request by Daniel-Lam**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/EUI-1970

### Change description ###
Implement S2S authentication as pluggable Express middleware.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: customTypes/otp/index.d.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+declare module 'otp'
```
Comment (ronaldmansveld) [inline customTypes/otp/index.d.ts:1]: Is this file still needed now you've switched to node-otp?

File: src/auth/s2s/s2s.class.spec.ts [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        // sinon.stub(serviceAuth, 'postS2SLease').resolves(token)
+        // eslint-disable-next-line @typescript-eslint/camelcase
+        // sinon.stub(jwt_decode.prototype).returns({ exp: Math.floor(Date.now() + 10000 / 1000) })
+    })
+
+    it('should be true that S2SAuth is defined', () => {
+        expect(S2SAuth).toBeDefined()
+    })
+
+    // xit('should return true that the cache is valid', () => {
```
Comment (ronaldmansveld) [inline src/auth/s2s/s2s.class.spec.ts:38]: xit already disables a test, no need to comment it out. If the test is no longer relevant, best to remove the code.

File: src/auth/s2s/s2s.class.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+export class S2SAuth extends events.EventEmitter {
+    protected readonly router = Router({ mergeParams: true })
+
+    protected s2sConfig!: S2SConfig
+
+    constructor() {
+        super()
+    }
+
+    public configure = (s2sConfig: S2SConfig): RequestHandler => {
```
Comment (ronaldmansveld) [inline src/auth/s2s/s2s.class.ts:26]: why this notation, and not simply
```
public configure(config: S2SConfig): RequestHandler { ... }
```

File: src/auth/s2s/s2s.class.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        this.s2sConfig = s2sConfig
+        this.router.use(this.s2sHandler)
+
+        return this.router
+    }
+
+    public s2sHandler = async (req: Request, res: Response, next: NextFunction): Promise<void> => {
```
Comment (ronaldmansveld) [inline src/auth/s2s/s2s.class.ts:33]: as above, but as `public async s2sHandler(...): Promise<void> { ... }`

File: src/auth/s2s/s2s.class.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                    return
+                }
+            }
+        } catch (error) {
+            logger.log('S2SAuth error:', error)
+            next(error)
+        }
+    }
+
+    public validateCache = (): boolean => {
```
Comment (ronaldmansveld) [inline src/auth/s2s/s2s.class.ts:57]: as above

File: src/auth/s2s/s2s.class.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+// TODO: To be replaced with a proper logging library
+const logger = console
+
+// Cache of S2S tokens, indexed by microservice name
+const cache: { [key: string]: S2SToken } = {}
+
+export class S2SAuth extends events.EventEmitter {
+    protected readonly router = Router({ mergeParams: true })
+
+    protected s2sConfig!: S2SConfig
```
Comment (ronaldmansveld) [inline src/auth/s2s/s2s.class.ts:20]: Why is this marked as not null (with the exclamation mark), but isn't it given a default value?

File: src/auth/s2s/s2s.class.ts [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+    public validateCache = (): boolean => {
+        logger.info('Validating S2S cache')
+        const currentTime = Math.floor(Date.now() / 1000)
+        if (!cache[this.s2sConfig.microservice]) {
+            return false
+        }
+        return currentTime < cache[this.s2sConfig.microservice].expiresAt
+    }
+
+    public getToken = (): S2SToken => {
```
Comment (ronaldmansveld) [inline src/auth/s2s/s2s.class.ts:66]: as above

File: src/auth/s2s/s2s.class.ts [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        return cache[this.s2sConfig.microservice]
+    }
+
+    public generateToken = async (): Promise<string> => {
```
Comment (ronaldmansveld) [inline src/auth/s2s/s2s.class.ts:70]: as above
also, should this method be public?

File: src/auth/s2s/s2s.class.ts [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+        cache[this.s2sConfig.microservice] = {
+            expiresAt: tokenData.exp,
+            token,
+        } as S2SToken
+
+        return token
+    }
+
+    public serviceTokenGenerator = async (): Promise<string> => {
```
Comment (ronaldmansveld) [inline src/auth/s2s/s2s.class.ts:84]: as above

File: src/auth/s2s/serviceAuth.spec.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+//             microservice: s2sConfig.microservice,
+//             oneTimePassword,
+//         })
+//         expect(response).to.equal('okay')
+//     })
+// })
+
+it('should serviceAuth', () => {
+    expect(true).toBeTruthy()
+})
```
Comment (ronaldmansveld) [inline src/auth/s2s/serviceAuth.spec.ts:43]: new spec file, but all is commented out apart from a non-test?

File: src/auth/s2s/serviceAuth.ts [hunk 11]
```diff
@@ -0,0 +1,19 @@
+import { totp } from 'node-otp'
+import { http } from '../../http/http'
+import { S2SConfig } from './s2sConfig.interface'
+
+// TODO: To be replaced with a proper logging library
+const logger = console
+
+export async function postS2SLease(s2sConfig: S2SConfig): Promise<string> {
```
Comment (ronaldmansveld) [inline src/auth/s2s/serviceAuth.ts:8]: Why is this is a separate exported function instead of part of the s2s class that's created?

File: src/auth/s2s/serviceAuth.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+    const oneTimePassword = totp({ secret: s2sConfig.s2sSecret })
+
+    logger.info('generating from secret: ', s2sConfig.s2sSecret, s2sConfig.microservice, oneTimePassword)
```
Comment (ronaldmansveld) [inline src/auth/s2s/serviceAuth.ts:11]: Don't think logging out secrets and passwords is a good idea for code to be checked in ;)

File: customTypes/otp/index.d.ts (hunk 1 above)
Comment (Daniel-Lam) [reply customTypes/otp/index.d.ts:1]: Good point, @ronaldmansveld - I knew I'd forget to remove this :)

File: src/auth/s2s/s2s.class.spec.ts (hunk 2 above)
Comment (Daniel-Lam) [reply src/auth/s2s/s2s.class.spec.ts:38]: Test was temporarily disabled using `xit` before being commented out as requested by @paddybasi, because it was using `sinon-express-mock` and hasn't been migrated to Jest yet.

File: src/auth/s2s/serviceAuth.spec.ts (hunk 10 above)
Comment (Daniel-Lam) [reply src/auth/s2s/serviceAuth.spec.ts:43]: All tests temporarily commented out at @paddybasi's request because they are using `sinon`/`chai`, etc. and are yet to be migrated to Jest.

File: customTypes/otp/index.d.ts (hunk 1 above)
Comment (Daniel-Lam) [reply customTypes/otp/index.d.ts:1]: Now removed

File: src/auth/s2s/s2s.class.ts (hunk 3 above)
Comment (Daniel-Lam) [reply src/auth/s2s/s2s.class.ts:26]: Amended accordingly

File: src/auth/s2s/s2s.class.ts (hunk 4 above)
Comment (Daniel-Lam) [reply src/auth/s2s/s2s.class.ts:33]: Amended accordingly

File: src/auth/s2s/s2s.class.ts (hunk 5 above)
Comment (Daniel-Lam) [reply src/auth/s2s/s2s.class.ts:57]: Amended accordingly

File: src/auth/s2s/s2s.class.ts (hunk 6 above)
Comment (Daniel-Lam) [reply src/auth/s2s/s2s.class.ts:20]: This has now been given a default, dummy config

File: src/auth/s2s/s2s.class.ts (hunk 7 above)
Comment (Daniel-Lam) [reply src/auth/s2s/s2s.class.ts:66]: Amended accordingly

File: src/auth/s2s/s2s.class.ts (hunk 8 above)
Comment (Daniel-Lam) [reply src/auth/s2s/s2s.class.ts:70]: Amended accordingly and made private

File: src/auth/s2s/s2s.class.ts (hunk 9 above)
Comment (Daniel-Lam) [reply src/auth/s2s/s2s.class.ts:84]: Amended accordingly

File: src/auth/s2s/serviceAuth.ts (hunk 12 above)
Comment (Daniel-Lam) [reply src/auth/s2s/serviceAuth.ts:11]: Agreed - now removed :)

File: src/auth/s2s/serviceAuth.ts (hunk 11 above)
Comment (Daniel-Lam) [reply src/auth/s2s/serviceAuth.ts:8]: This function has now been moved into the main `s2s.class.ts` file.
